### PR TITLE
Add functionalities to get,update, delete facility by id (user), making facility registration request by facility user, and admin handling of those requests

### PR DESCRIPTION
**Pull request by nurgissas**
## What was added
- Add get, update, delete facility by user id methods to userServices
- Corresponding to the above methods changes in controller, router
- Add functionality to request facility registration by facility user (methods, controllers, routes)
- Add admin handling of those requests
- Admin can create facility with request.content
- Admin can get (by authorId or all), delete facility registration request
- Last Edit: bug fix of decline facility registration request

## Related Issues
- Didn't run unit test with existing facility id 
- Tested all facility-registration-request related tests
- But run negative unit tests to endpoints targetting above methods

## Future Works
- Search engine to map services

## Checklist
- [x] I have performed unit tests
- [ ] I have added necessary docs
- [ ] I have checked for potential bugs & issues

### CODE REVIEW

**Comment by cwwojin** [comment]
Great work!
1. `POST /api/facilities/facility-requests` - the `content` should have required field `address` and optional fields `openingHours, menu, preferences, stampRuleset, stampRewards` that include all data to be inserted. validate fields accordingly
2. `GET /api/admin/facility-requests` - OK, but adding optional query parameters `authorId, status` would be nice
3. `GET /api/admin/facility-requests/:id` - OK
4. `POST /api/admin/facility-requests/:id/accept` - insert `preferences, stampRuleset, stampRewards` also
5. `POST /api/admin/facility-requests/:id/decline` - OK
6. `DELETE /api/admin/facility-requests/:id` - OK

Also, you don't need to update `updated_at = now()` -> this will be done by DB

File: myapp/src/services/adminService.js [hunk 1]
```diff
@@ -1,104 +1,166 @@
...
+    return result.rows;
+  },
+  /** accept a facility registration request */
+  acceptFacilityRegistrationRequest: async (id, adminId) => {
+    const request = await module.exports.getFacilityRegistrationRequest(id);
+    if (!request || request.status !== 0) {
+      throw { status: 404, message: `Request doesn't exist or is not pending` };
+    }
+
+    const data = JSON.parse(request.content);
```
Comment (cwwojin) [inline myapp/src/services/adminService.js]: This will give error if the `content` is an object and not JSONstring

**Review by cwwojin — CHANGES_REQUESTED** [review]

File: myapp/src/services/adminService.js [hunk 2]
```diff
@@ -1,104 +1,166 @@
...
+      // Create the facility and related entries
+      const facility = await facilityService.createFacility(data);
+
+      // Update the registration request status
+      const query = {
+        text: `UPDATE facility_registration_request SET status = $1, respond_date = NOW(), updated_at = NOW() WHERE id = $2 RETURNING *`,
+        values: [1, id],
+      };
+      const result = await db.query(query);
+
```
Comment (cwwojin) [inline myapp/src/services/adminService.js:142]: Here, also insert the facility-manager relationship
`"insert into manages (user_id, facility_id) values ($1, $2)", [request.author_id, facility.id]`

File: myapp/src/services/facilityService.js [hunk 3]
```diff
@@ -167,6 +165,14 @@ class FacilityService {
   }
 
   async updateFacility(id, data) {
+    const facility = await this.getFacilityById(id);
+    const status = await this.getFacilityRegistrationStatus(facility.authorId);
```
Comment (cwwojin) [inline myapp/src/services/facilityService.js]: This check will not allow facility update, if the facility is either
- made by admin without the request process, or any facilities that are pre-existent in DB
- request is deleted after being accepted

**Review by cwwojin — CHANGES_REQUESTED** [review]

**Comment by nurgissas** [comment]
POST /api/facilities/facility-requests:
- Added optional fields: address, openingHours, menu, preferences, stampRuleset, stampRewards.
- Validated as objects or arrays.
- Content is treated as object, not string

GET /api/admin/facility-requests:
- Added optional query parameters: authorId, status.

POST /api/admin/facility-requests/:id/accept:
- Added insertion for preferences, stampRuleset, stampRewards.

**Comment by cwwojin** [comment]
@nurgissas 
My changes :

1. `GET /api/admin/facility-requests` change query parameter name to `authorId -> user` (for consistency with other method)
2. Edit format of `content` in facility request : `stampRuleset: { totalCnt: 10, rewards: []}`
    stampRuleset.rewards is the array with rewards -> to use `facilityService.createStampRuleset` for inserting all
1. Change most fields that use `notEmpty() -> exists().isString()` except for `businessId` Also, allow facility URL and email to be empty string

**Comment by cwwojin** [comment]
Great Work ! I will merge this after solving conflict with dev branch

**Review by cwwojin — APPROVED** [review]

**Pull request merged**